pkg/mcp/httpclient.go

```diff
@@ -77,29 +86,50 @@ func (s *HTTPClient) newRequest(ctx context.Context, method string, in any) (*ht
 		// Don't add because some *cough* CloudFront *cough* proxies don't like it
 		req.Header.Set("Accept", "application/json, text/event-stream")
 	}
+
+	if body != nil {
+		req.Header.Set("Content-Type", "application/json")
+	}
 	return req, nil
 }
 
-func (s *HTTPClient) startSSE(ctx context.Context, msg *Message) error {
+func (s *HTTPClient) startSSE(ctx context.Context, msg *Message, lastEventID any) error {
 	gotResponse := make(chan error, 1)
 	req, err := s.newRequest(ctx, http.MethodGet, nil)
 	if err != nil {
 		return err
 	}
 
+	if lastEventID != nil {
+		req.Header.Set("Last-Event-Id", fmt.Sprintf("%v", lastEventID))
+	}
+
 	resp, err := http.DefaultClient.Do(req)
 	if err != nil {
 		return err
 	}
 
 	if resp.StatusCode != http.StatusOK {
 		_ = resp.Body.Close()
+		// If msg is nil, then this is an SSE request for HTTP streaming.
+		// If the server doesn't support a separate SSE endpoint, then we can just return.
+		if msg == nil && resp.StatusCode == http.StatusMethodNotAllowed {
+			return nil
+		}
 		return fmt.Errorf("failed to connect to SSE server: %s", resp.Status)
 	}
 
-	go func() {
-		defer s.waiter.Close()
-		defer resp.Body.Close()
+	go func() (err error, send bool) {
+		defer func() {
+			if err != nil {
+				s.waiter.Close()
+				if send {
+					gotResponse <- err
+				}
+			}
```

Comment:
This is the only weird part. We need to close the waiter if we're done (indicated by an error here), but keep it open if we are going to retry.